typing/typecore.ml

```diff
@@ -3284,11 +3269,15 @@ let type_approx_fun_one_param env label spato ty_expected ~first ~in_function =
       in
       raise (Error(loc_fun, env, err))
   in
-  if has_poly then begin
+  let () =
     match spato with
     | None -> ()
-    | Some spat -> type_pattern_approx env spat ty_arg
-  end;
+    | Some spat ->
+      if has_poly || not (Btype.tpoly_is_mono ty_arg) then
+        type_pattern_approx env spat ty_arg
+      else
+        type_pattern_approx env spat (Btype.tpoly_get_mono ty_arg)
+  in
```

Comment:
This smells weird to me. What are these branches supposed to do? Could we do the same by checking for the case `Ptyp_poly(ty, [])` in `approx_type`?

Reply:
I agree that this looks off. I think the old code from my PR was trying to only look at type annotations when they were polymorphic so as not to change the existing behaviour for other annotations. Now that #12315 has made it look at all annotations, we probably don't need to distinguish the polymorphic case so much.

Reply:
I indeed added the second branch to fix the tests introduced in #12315.
At first I simply tried to remove the `has_poly` condition but it doesn't work without unwrapping the tpoly.

Reply:
Looking at what `type_function` does, you should probably be doing this logic here: https://github.com/ocaml/ocaml/blob/95c5177ef97bdb6524273297ca27cf853febdac0/typing/typecore.ml#L5202